### ISSUE TITLE
importPage() parameter are missleading and fault-prone

### ISSUE THREAD

**Issue by JanSlabon** [mpdf/mpdf#480]
We're currently preparing a new version of FPDI for the public (https://github.com/Setasign/FPDI/releases/tag/v2.0.0-beta2).

In the near future we would like to make mPDF compatible with this version of FPDI but actually I'm of two minds with the parameter from importPage():

    function ImportPage($pageno = 1, $crop_x = null, $crop_y = null, $crop_w = 0, $crop_h = 0, $boxName = '/CropBox')

I understand what Ian would like to accomplish with the $crop_* parameters but he didn't thought about boundaries which do not have their origin on [0,0] or a page that is rotated through its Rotate entry in the page dictionary. An example document is here: [[-1000 -1000 -500 -500].pdf](https://github.com/mpdf/mpdf/files/1268392/-1000.-1000.-500.-500.pdf)

The $crop_* parameters are replacement values for the page boundary values of the imported page and do not crop the imported page directly. To accomplish this it would be required to calculate the relative values to the imported page boundary. Also the rotation of the imported page shall be taken into account... after all this is much effort for a feature that can be done much easier in the graphic state of a page, before the imported page is used (such as by _setClippingPath()). Over all we never got a request for such feature on the offical FPDI version for more than a decade.

Would it be ok, to remove the $crop_* parameter in an upcoming pull request or will this be a no-go?

**Comment by finwe** [comment]
How about introducing a new method and marking ImportPage as deprecated until v 8? Shame to lose the method name, though.

**Comment by JanSlabon** [comment]
A new method name would be somewhat odd.

Isn't a new major version (7) made for such changes? I mean by removing the parameters we remove a faulty behavior. Overall I guess that this "feature" was not used very often.

After all it would be possible to keep the behavior but as it is wrong it needs some overcoming at our end :-)

**Comment by finwe** [comment]
Sure, we _can_ introduce breaking changes in major version per semver "spec", however, I would like the new method to be error-prone and to warn when being used the old way so that people upgrading know when they missed an occurence in their code. 

How would you propose to do that? Throwing an error/exception when there is a second parameter that is not a string? when there are more than two parameters? All these solutions  are not 100% foolproof. I'd rather not introduce another fault-prone state by (insufficiently) fixing another.

**Comment by JanSlabon** [comment]
Isn't this enough:

    if ($box === null || is_numeric($box)) {
        throw new \InvalidArgumentException('...');
    }

?

Later the $box parameter would be checked for a valid string value anyway.

Furthermore a good IDE will also mark the method call if there are more arguments passed than accepted.

Anyhow to offer the cropping/clipping feature we need a public method for clipping. Is there something implemented or is it only the _setClippingPath() method (which looks like a protected one due to the underscore)? Did I missed something?

**Comment by finwe** [comment]
I don't know about any public clipping method. Yes, underscored methods were likely meant to be private, however it often seems to me that Ian himself was not entirely sure about naming and visibility.

**Comment by finwe** [comment]
OK, your proposed solution to removing parameters seems OK to me, lets solve the little issues over the PR.

Please, introduce a `\Mpdf\Exception\InvalidArgumentException` class extending `\Mpdf\MpdfException` for more fine-grained exception handling.

**Comment by finwe** [comment]
@JanSlabon Have you come up with something? The last blocking error in 7.0 has been fixed and I'd like to release a RC soon. This should be done before that.

**Comment by JanSlabon** [comment]
@finwe Sorry but we hadn't had the time to start this until now. I also guess that this will not be done within the next 3 or 4 weeks due to other higher priority projects and holidays, sorry.

**Comment by finwe** [comment]
I understand. Can we at least finalize the method signature? The underlying changes could be done afterwards, I guess.

**Comment by finwe** [comment]
I have looked at the code and at the examples where the crop functionality is actually used and decided to postpone this until there is a working replacement. Finalizing this would only postpone 7.0 which is now otherwise ready for a RC.

Thanks for the proposition, I'll be happy to review it in the development branch for the next major version.

**Comment by JanSlabon** [comment]
@finwe I'm sorry but actually I have simply no time... 7.1 should work, too ;-)

**Comment by drewwynne0** [comment]
+1 - @JanSlabon @finwe  Does this look like something that could happen sooner rather than later? We are currently utilising FPDI ^2.0 and would rather not have to role back to be able to utilise mPDF.

**Comment by JanSlabon** [comment]
Sorry but we've actually no capacities for this, sorry. It's on the list but other tasks simply have a higher priority.

**Comment by drewwynne0** [comment]
@JanSlabon @finwe Am I correct in saying that this will be resolved as part of https://github.com/mpdf/mpdf/issues/578 - and is therefore under the 2-3 weeks mentioned here 11 days ago - https://github.com/mpdf/mpdf/issues/871#issuecomment-425846450 ? 🤞

**Comment by JanSlabon** [comment]
I hope we will have some time to do this within the next week, yes. Anyhow, this is unpaid work and we may have to change the priority again, so for sure we do not give any guarantee on any schedule.

**Comment by drewwynne0** [comment]
@JanSlabon completely understand!  Really appreciate your (and all others) contributions to this and related projects for the community!  Thanks for the update!

**Comment by JanSlabon** [comment]
I'm on it: https://github.com/JanSlabon/mpdf/tree/FPDI-2-Update

**Comment by finwe** [comment]
Solved in FPDI 2 integration.
